### DISCUSSION TITLE
Cracking from phone

### DISCUSSION THREAD

**Discussion by Stepzor11** [AloneLiberty/FlipperNested#2] [Ideas]
As a mere suggestion, consider to develop a script to run on Android and maybe iOS to calculate the keys (maybe after a manual import of the nonces from the  Fipper's SD).
It would increase greatly the portability of the attack

**Comment by Suzanskyi** [comment]
Absolutely great idea. Currently we have to have notebook right in hand in order to crack card. But thank you for this opportunity as well!

**Comment by AloneLiberty** [comment]
Some info:

**Android**:
Script should work on Termux with version 2.2.2+, you can recover keys from files via phone.
Termux doesn't provide any serial ports (https://github.com/termux/termux-packages/issues/4618#issuecomment-561047747), only libusb access. It means full rewrite of bridge to support libusb. Same goes for TCP forwarding.
**iOS**:
iSH emulates x86, I think it's pretty self-explanatory.
![image](https://user-images.githubusercontent.com/111039319/236019769-def56777-7cb8-4f57-b4dd-4eca930ba9a0.png)

**Reply by Stepzor11** [reply]
That's why I suggested to let the user do the import/export manually.

The FAP could save the data in a specific folder, so the user can import them (via SD, USB or whatever he likes) in a specific folder on the phone, run the script and import the obtained keys manually in the Flipper user dictionary.

That should need a way lower effort than rewriting the bridge, on Android at least.

I am pretty sure many people would love that, me first 😁

**Reply by AloneLiberty** [reply]
You might check FAQ and --help to FlipperNestedRecovery, it's been around for a while :wink:

**Reply by Stepzor11** [reply]
Thanks a lot!

Few questions, if you 'd like to answer:

Is it mandatory to keep the flipper connected the whole time?
What happens if it disconnects?

Screen lock on flipper seems like a disconnection, can you confirm?

Is it possible to run the script on a local file on the pc and import the result manually on flipper (to avoid the need of phisical connection, and maybe use a remote machine)?
It seems like with --file it could work, but if you are so kind to explain it better, especially what to do after the recovery, I'd really appreciate

**Reply by AloneLiberty** [reply]
If Flipper disconnects in recovery process - your keys will be saved in current directory, you will need to copy them into /ext/nfc/.nested/ directory manually (should add this to FAQ)

**Reply by AloneLiberty** [reply]
--file is actually only for debugging purpose, in any other case you should use real connection (for HardNested you must copy whole folder and etc)

**Reply by AloneLiberty** [reply]
After --file process is same as with disconnected Flipper (but in some cases you will need to fix file name, should be UID.keys, not UID.nonces.keys or others)

**Comment by Stepzor11** [comment]
I know I might sound boring, thanks for your patience.
Which is the "current directory" where I'll find them?

If I run the script on a local file ("--file" oprtion), which one should I choose?
I see that in .nested folder there is a UID.nonces file and a UID folder, containing 0_0, 0_1, ect. nonces files

**Reply by AloneLiberty** [reply]
If case of HardNested you should copy entirely .nested to PC and run --file UID.nonces (and now I will confuse you: you can run even on 0_0.nonces, but will recover only one key, UID.nonces will recover all)

Current directory is where are you running FlipperNested

**Reply by Stepzor11** [reply]
So if I run it on UID.nonces it will look also for the files in UID folder, am I right?

**Reply by AloneLiberty** [reply]
Yes, they are required

**Reply by Stepzor11** [reply]
Great!
Even if it should be used only for debug, --file could give me the ability to use the script on a remote machine.
Could it be a good option?

While we are talking, I am running it and for what I can see it started a bit slowly, compared to starting it from Flipper, but it seems to work

**Reply by AloneLiberty** [reply]
It gives flexibility with some restrictions, so not suitable for everyone (like regular 1 click script). Debug only is probably wrong words

**Reply by Stepzor11** [reply]
Well, I'm a bit ignorant in scripting and full of questions, but I'm a quick learner.
Thanks again for your help ❤️

The time expectation seems roughly twice as the one expected when it was running from flipper's file.
Is there a specific reason or it is just a different esteem?

So they do the exact same thing, just that --file is not a "1 click script" or do they have some sort of differences that explain a slower time?

**Reply by AloneLiberty** [reply]
They even share same file text parser 😉 
So they will work same (some time difference may happen, CPU load or something in HardNested)

**Reply by Stepzor11** [reply]
Thanks a lot!

No more question right now ❤️

**Reply by Stepzor11** [reply]
Works greatly! ❤️

**Reply by Stepzor11** [reply]
Other suggestions:

it would be pretty useful to have the ability to stop the script, making it save the keys calculated until the stop.
Calculating all the keys might be a waste of time, plus you can deduct the other keys based on some sort of logic in the key generation (for example in a case I tested part of them are sequential).

Also being able to pause it, might be useful.

I hope that my thoughts can be useful.

Thanks for your great work!

**Reply by AloneLiberty** [reply]
I'd say pausing is impossible, but saving keys if stopped manually is great idea, I will implement it

**Reply by AloneLiberty** [reply]
Added in 2.3.1 version of FlipperNestedRecovery

**Comment by Stepzor11** [comment]
Happy to help 😁
________________________________
